master/static/migrations/20240108170651_add-projects-trigger.tx.up.sql

```diff
@@ -0,0 +1,70 @@
+ALTER TABLE projects ADD COLUMN IF NOT EXISTS seq bigint;
```

Comment:
note: would it be worthwhile to instead use a timestamp, instead of creating redundant sequences to populate the seq value?

Reply:
I don't think so - we just care about last modified/order of modification right?

Reply:
The more I'm thinking about how we're using seq, it seems like a timestamp would be more than sufficient, there's pros & cons to both approaches

**Pros of using seq:**
- Reduces problem scope to only monitoring for order of modifications
- time independent
- reduces memory footprint in early use

**Cons of seq:**
- Adds additional sequence that must be created in order to track changes
- Potential for misleading information when comparing across tables (seq as a concepts should only be used in reference to a single table, and can't be use to order events between tables)
- Potential overflow issues in system with a large volume of changes

**Pros of timestamp:**
- standard use case of a timestamp (something like `last_modified_at`)
- can be used demonstrate timeline of changes between tables 
- Better UX if a streaming client is interested in subscribing to events after a specific time
- Avoids overflow problem 
- can be represented as an integer for operations & storage
 
**Cons of timestamp:**
- Potential races & timestamp inconsistencies; however this would be a similar problem with seq just not as granular

Reply:
Ultimately, a timestamp would not only enforce strict ordering, but also provide a more intuitive value that could be used in startup messages without having to have prior context on the system. 

I've been debating between, using a timestamp, using a hybrid of a timestamp and a cycled sequence, or just a sequence value; and I believe a timestamp would provide the preferred behavior without the downsides of overflow and redundant sequences.

Reply:
I think doing this based on timestamps introduces the possibility of bugs for little or no upside.  The most important thing from the streaming updates design perspective is a guaranteed monotonically increasing sequence.  Time seems like it would do the trick but if a server's clock ever got ahead (and subsequently corrected backwards) it could create situations where clients miss data.  This creates a system that is "usually" bug-free, but eventually will exhibit some buggy behavior due to some customer's server having a faulty clock.

On the other hand, postgres has a tool to solve this exact problem in a reliable way, and all we have to do is use it.

Regarding the ability to subscribe to updates since a particular point in time... I don't think that streaming updates is going to be end-user facing.  There may be end-user facing features which depend on streaming updates (the webui, an event monitoring system in the python sdk), but by far these features are of the "download what's there then stream updates" sort, not the "tell me about updates since time T" sort.

In the relatively rare cases where the client would indeed want to subscribe by time, like viewing recent master logs or something, those endpoints should track time separately.  The time column would be used just for time-based requests, while the sequence would be used for reliable streaming updates.

I'm curious what you mean by "redundant sequences"?  Or what the storage cost concern is between a sequence and a timestamp?  They should both be int8 columns in postgres.  And int8 won't overflow any time soon, even timestamps are stored in 8 bytes.

Reply:
Originally, there had been concerns about having a sequence for each table; however, in order to ensure a modification on one table doesn't block a modification of another, having separate sequences makes sense.

I had also thought we had been using integer for the sequence, which was where my concern had been (not int8/bigint)